File: src/main/java/com/footverse/order/dto/CouponPreviewResponse.java
package com.footverse.order.dto;

import java.math.BigDecimal;

/**
 * The full server-computed checkout summary for the previewed items (dto-spec §14). Every amount is
 * computed by the server; the client never computes discounts or totals. The {@code code} / {@code
 * name} are present only when a valid coupon was supplied.
 *
 * @param code           the applied coupon code, if a valid coupon was supplied
 * @param name           the applied coupon name, if a valid coupon was supplied
 * @param subtotal       the sum of the selected line totals
 * @param discountAmount the computed discount (0 when no coupon)
 * @param shippingFee    the flat shipping fee
 * @param total          {@code subtotal − discountAmount + shippingFee}
 */
public record CouponPreviewResponse(
        String code,
        String name,
        BigDecimal subtotal,
        BigDecimal discountAmount,
        BigDecimal shippingFee,
        BigDecimal total) {
}

File: app/lib/features/order/models/coupon_preview_response.dart
import 'package:json_annotation/json_annotation.dart';

import '../../product/models/decimal_json.dart';

part 'coupon_preview_response.g.dart';

/// The full server-computed checkout summary for the previewed items
/// (dto-spec §14).
///
/// Every amount is computed by the server and rendered exactly as delivered; the
/// client never computes discounts or totals (dto-spec §1; business-rules →
/// Checkout Preview). [code] / [name] are present only when a valid coupon was
/// supplied.
@JsonSerializable(createToJson: false)
class CouponPreviewResponse {
  const CouponPreviewResponse({
    required this.subtotal,
    required this.discountAmount,
    required this.shippingFee,
    required this.total,
    this.code,
    this.name,
  });

  factory CouponPreviewResponse.fromJson(Map<String, dynamic> json) =>
      _$CouponPreviewResponseFromJson(json);

  final String? code;
  final String? name;
  @JsonKey(fromJson: decimalFromJson)
  final double subtotal;
  @JsonKey(fromJson: decimalFromJson)
  final double discountAmount;
  @JsonKey(fromJson: decimalFromJson)
  final double shippingFee;
  @JsonKey(fromJson: decimalFromJson)
  final double total;
}
